Install MQTT connection tool for Python.

In [35]:
pip install paho-mqtt

Note: you may need to restart the kernel to use updated packages.


Subscribe to the relevant topic which publishes the real time data of EK-220 electronic volume corrector.
    
Create a dictionary of parameters depending on the data received and append the message payload to the dictionary appropriately.
    
Interrupt the kernel to finish your test when enough amount of data is received.

In [1]:
import paho.mqtt.client as mqtt
from datetime import datetime

my_dict = {'Vaxt (ISO format)':[],
           'Vaxt (real)':[],
           'Mexaniki göstərici - VmT(m3)':[],
           'Elektron göstərici - VbT(m3)':[],
           'Ölçmə şəraitində sərfiyyat - Qm(m3/h)':[],
           'Baza şəraitində sərfiyyat - Qb(m3/h)':[],
           'Təzyiq - p(bar)':[],
           'Temperatur - T(C)':[],
           'Korreksiya əmsalı - C':[],
           'Qazın sıxılma əmsalı - K':[],
           'Baza şəraitində xüsusi çəki - Rhob(kg/m3)':[],
           'VmD - VmD(m3)':[],
           'VbD - VbD(m3)':[]}

def on_connect(client, userdata, flags, rc):
    print("Connected with result code " + str(rc))
    client.subscribe('ek220', qos = 0)

def on_log(client, userdata, level, buff):
    print(buff)
    
def on_message(client, userdata, msg):
    print(msg.topic + " " + str(msg.payload))
    
    string = str(msg.payload)
    string = string[3:len(string)-2]   
    my_list = string.split(',')  
    now = datetime.now()
    current_time = now.strftime("%d.%m.%Y %H:%M:%S")

    my_dict['Vaxt (ISO format)'].append(now)
    my_dict['Vaxt (real)'].append(current_time)
    for i in range(len(my_list)):
        if i == 1:
            if my_list[i][5:len(my_list[i])-1] == '':
                my_dict['Mexaniki göstərici - VmT(m3)'].append(None)
            else:
                my_dict['Mexaniki göstərici - VmT(m3)'].append(float(my_list[i][5:len(my_list[i])-1]))
        elif i == 2:
            if my_list[i][5:len(my_list[i])-1] == '':
                my_dict['VmD - VmD(m3)'].append(None)
            else:
                my_dict['VmD - VmD(m3)'].append(float(my_list[i][5:len(my_list[i])-1]))
        elif i == 3:
            if my_list[i][5:len(my_list[i])-1] == '':
                my_dict['Elektron göstərici - VbT(m3)'].append(None)
            else:
                my_dict['Elektron göstərici - VbT(m3)'].append(float(my_list[i][5:len(my_list[i])-1]))
        elif i == 4:
            if my_list[i][5:len(my_list[i])-1] == '':
                my_dict['VbD - VbD(m3)'].append(None)
            else:
                my_dict['VbD - VbD(m3)'].append(float(my_list[i][5:len(my_list[i])-1]))
        elif i == 6:
            if my_list[i][5:len(my_list[i])-1] == '':
                my_dict['Ölçmə şəraitində sərfiyyat - Qm(m3/h)'].append(None)
            else:
                my_dict['Ölçmə şəraitində sərfiyyat - Qm(m3/h)'].append(float(my_list[i][5:len(my_list[i])-1]))
        elif i == 7:
            if my_list[i][5:len(my_list[i])-1] == '':
                my_dict['Baza şəraitində sərfiyyat - Qb(m3/h)'].append(None)
            else:
                my_dict['Baza şəraitində sərfiyyat - Qb(m3/h)'].append(float(my_list[i][5:len(my_list[i])-1]))
        elif i == 8:
            if my_list[i][5:len(my_list[i])-1] == '':
                my_dict['Təzyiq - p(bar)'].append(None)
            else:
                my_dict['Təzyiq - p(bar)'].append(float(my_list[i][5:len(my_list[i])-1]))
        elif i == 9:
            if my_list[i][6:len(my_list[i])-1] == '':
                my_dict['Temperatur - T(C)'].append(None)
            else:
                my_dict['Temperatur - T(C)'].append(float(my_list[i][6:len(my_list[i])-1]))
        elif i == 10:
            if my_list[i][6:len(my_list[i])-1] == '':
                my_dict['Korreksiya əmsalı - C'].append(None)
            else:
                my_dict['Korreksiya əmsalı - C'].append(float(my_list[i][6:len(my_list[i])-1]))
        elif i == 11:
            if my_list[i][6:len(my_list[i])-1] == '':
                my_dict['Qazın sıxılma əmsalı - K'].append(None)
            else:
                my_dict['Qazın sıxılma əmsalı - K'].append(float(my_list[i][6:len(my_list[i])-1]))
        elif i == 12:
            if my_list[i][6:len(my_list[i])-1] == '':
                my_dict['Baza şəraitində xüsusi çəki - Rhob(kg/m3)'].append(None)
            else:
                my_dict['Baza şəraitində xüsusi çəki - Rhob(kg/m3)'].append(float(my_list[i][6:len(my_list[i])-1]))
    
client = mqtt.Client("Erturk")
client.username_pw_set(username = "***", password = "***")
client.on_log = on_log
client.on_connect = on_connect
client.on_message = on_message

print("Connecting to broker.")
try:
    client.connect(host = "***.***.***.***", port = 1883, keepalive = 60, bind_address = "")
except:
    print("Connection failed!")
    exit(1)

client.loop_forever()

Connecting to broker.
Sending CONNECT (u1, p1, wr0, wq0, wf0, c1, k60) client_id=b'Erturk'
Received CONNACK (0, 0)
Connected with result code 0
Sending SUBSCRIBE (d0, m1) [(b'ek220', 0)]
Received SUBACK
Received PUBLISH (d0, q0, r0, m0), 'ek220', ...  (436 bytes)
ek220 b'{"1":"12.04.22.20 11:59:40","2":"190507.400","3":"0.0","4":"194476.7196","5":"0.0","6":"24","7":"52.3000","8":"53.4000","9":"1.0370","10":"20.7200","11":"1.0210","12":"0.9999","13":"0.8313","14":"Nerimanov","15":"0","16":"0","17":"0","18":"0","19":"0","20":"0","21":"0","22":"0","23":"0","24":"0","25":"0","26":"45056","27":"4096","28":"40960","29":"0","30":"32768","31":"8192","32":"8192","33":"0","34":"0","35":"0","36":"0","37":"30"}'
Received PUBLISH (d0, q0, r0, m0), 'ek220', ...  (436 bytes)
ek220 b'{"1":"12.04.22.20 12:00:02","2":"190507.3400","3":"0.0","4":"194477.269","5":"0.0","6":"24","7":"52.8000","8":"54.0000","9":"1.0392","10":"20.7800","11":"1.0229","12":"1.0000","13":"0.8313","14":"Nerimanov","15":"0","16":

KeyboardInterrupt: 

Check if there is any problem while receiving data.

One of the anomalies that can occur during this process may be related to the timestamp data because one of the power source of EK-220 devices is solar panel. So, if solar panel stops working, GSM modem also stops working. Thus, we can understand at which time modem did not work appropriately from timestamp data.

In [38]:
from decimal import Decimal

for i in range(len(my_dict['Vaxt (ISO format)']) - 1):
    difference = my_dict['Vaxt (ISO format)'][i + 1] - my_dict['Vaxt (ISO format)'][i]
    total_seconds = float(difference.total_seconds())
    hours, remainder = divmod(total_seconds, 60 * 60)
    minutes, seconds = divmod(remainder, 60)
    if int(hours) == 0 and int(minutes) == 0:
        print('{} saniyə'.format(round(Decimal(seconds), 2)))
    elif int(hours) == 0:
        print('{} dəqiqə {} saniyə'.format(int(minutes), round(Decimal(seconds), 2)))
    else:
        print('{} saat {} dəqiqə {} saniyə'.format(int(hours), int(minutes), round(Decimal(seconds), 2)))

21.96 saniyə
1 dəqiqə 48.54 saniyə
21.51 saniyə
21.78 saniyə
21.60 saniyə
21.55 saniyə
21.54 saniyə
21.55 saniyə
21.85 saniyə
21.89 saniyə
22.02 saniyə
22.02 saniyə
22.61 saniyə
23.51 saniyə


Another anomaly that can occur during this process may be related to the 'NULL' data for any parameter because when EK-220 device is working, it must send logical value. So, if it does not send any value, it means that corrector stopped working. Thus, we can understand at which time corrector did not work appropriately from any parameter.

In [39]:
for i in range(len(my_dict['Temperatur - T(C)'])):
    if my_dict['Temperatur - T(C)'][i] is None:
        print(f"There is problem in corrector data at time {my_dict['Vaxt (real)'][i]}")
    elif i == len(my_dict['Temperatur - T(C)']) - 1:
        print('No problem!')

No problem!


Since the output of timestamp data is 1 minute 48.54 seconds in one instance, it is almost one and a half minute lag.

So, let's check whether there is any problem in that period.

In [34]:
for key, value in my_dict.items():
    if key not in ('Vaxt (ISO format)', 'Vaxt (real)'):
        print(key, ':', value[0:3])

Mexaniki göstərici - VmT(m3) : [190507.4, 190507.34, 190507.64]
Elektron göstərici - VbT(m3) : [194476.7196, 194477.269, 194477.3353]
Ölçmə şəraitində sərfiyyat - Qm(m3/h) : [52.3, 52.8, 4.0]
Baza şəraitində sərfiyyat - Qb(m3/h) : [53.4, 54.0, 4.15]
Təzyiq - p(bar) : [1.037, 1.0392, 1.0549]
Temperatur - T(C) : [20.72, 20.78, 20.87]
Korreksiya əmsalı - C : [1.021, 1.0229, 1.0382]
Qazın sıxılma əmsalı - K : [0.9999, 1.0, 0.9999]
Baza şəraitində xüsusi çəki - Rhob(kg/m3) : [0.8313, 0.8313, 0.8313]
VmD - VmD(m3) : [0.0, 0.0, 0.0]
VbD - VbD(m3) : [0.0, 0.0, 0.0]
